### PR TITLE
Gh 321 blockchain api

### PR DESCRIPTION
**Pull request by chb0github**
A clean PR to main so we can discuss before showing to the team

### CODE REVIEW

File: src/blockchain/src/notifier.rs [hunk 1]
```diff
@@ -7,15 +7,16 @@ use crate::blockchain::Blockchain;
...
-pub type BlockchainResult = std::result::Result<(), BlockchainError>;
-pub type OnTransactionSettled = Box<dyn FnOnce(Transaction)>;
-pub type OnBlockEvent = Box<dyn FnMut(Block)>;
-
 // But we require certain bounds to get things done...
 impl Blockchain {
     // should we borrow or own this transaction?
-    pub fn submit_transaction(&mut self, trans: Transaction, on_done: OnTransactionSettled) {
-        self.trans_observers.insert(trans, on_done);
+    pub fn submit_transaction<CB: 'static + FnOnce(Transaction) + FnOnce(Transaction)>(
```
Comment (chb0github) [inline src/blockchain/src/notifier.rs]: I dont' think `FnOnce(Transaction)` twice is needed. I will try to remove the extra

Comment (prince-chrismc) [reply src/blockchain/src/notifier.rs]: Name is a little confusing

```suggestion
    pub fn submit_transaction<callback: 'static + FnOnce(Transaction) + FnOnce(Transaction)>(
```

File: src/blockchain/src/notifier.rs [hunk 2]
```diff
@@ -27,8 +28,11 @@ impl Blockchain {
         }
     }
 
-    pub fn add_block_listener(&mut self, on_block: OnBlockEvent) -> &mut Self {
-        self.block_observers.push(on_block);
+    pub fn add_block_listener<CB: 'static + FnMut(Block) + FnMut(Block)>(
```
Comment (prince-chrismc) [inline src/blockchain/src/notifier.rs]:
```suggestion
    pub fn add_block_listener<callback: 'static + FnMut(Block) + FnMut(Block)>(
```

File: src/blockchain/src/notifier.rs [hunk 3]
```diff
@@ -92,48 +97,51 @@ mod test {
...
         );
-        let mut called:  bool = false;
-        let mut lambda = |trans: Transaction| {
-            assert_eq!(transaction, trans);
-            called = true;
-        };
-        chain.submit_transaction(transaction.clone(), Box::new(lambda));
-        chain.notify_transaction_settled(transaction.clone());
-        assert!(called);
+        let called = Rc::new(Cell::new(false));
```
Comment (prince-chrismc) [inline src/blockchain/src/notifier.rs]: Can you add a comment where this magic came from

**Pull request closed without merging**